### PR TITLE
Add paddle.linalg.matrix_power OP

### PR DESCRIPTION
**Pull request by haolin-nju**
### PR types

New features

### PR changes

OPs

### Describe

This PR supports matrix_power calculation for linear algorithm module of Paddle. One may call `paddle.linalg.matrix_power` or `paddle.matrix_power` to use it.

An example is listed below, which is a python3 script:
```python
import paddle

x = paddle.to_tensor([[1, 2, 3],
                      [1, 4, 9],
                      [1, 8, 27]], dtype='float64')
print(paddle.matrix_power(x, 2))
# [[6.  , 34. , 102.],
#  [14. , 90. , 282.],
#  [36. , 250., 804.]]

print(paddle.matrix_power(x, 0))
# [[1., 0., 0.],
#  [0., 1., 0.],
#  [0., 0., 1.]]

print(paddle.matrix_power(x, -2))
# [[ 12.91666667, -12.75000000,  2.83333333 ],
#  [-7.66666667 ,  8.         , -1.83333333 ],
#  [ 1.80555556 , -1.91666667 ,  0.44444444 ]]
```

A detailed doc PR in Chinese is now available at [docs PR3727](https://github.com/PaddlePaddle/docs/pull/3727). Its preview version is available at [matrix_power doc preview](http://10.136.157.23:8090/documentation/docs/zh/api/paddle/matrix_power_cn.html?reviewVersion=jenkins-doc-review-2-195) before UTC+8 20/08/2021.

### CODE REVIEW

File: python/paddle/tensor/linalg.py [hunk 1]
```diff
@@ -941,3 +941,55 @@ def __check_input(x, vec):
         type='mv', inputs={'X': x,
                            'Vec': vec}, outputs={'Out': out})
     return out
+
+
+def matrix_power(x, n):
```
Comment (jzhang533) [inline python/paddle/tensor/linalg.py]: 缺少name参数

File: python/paddle/tensor/linalg.py [hunk 2]
```diff
@@ -941,3 +941,55 @@ def __check_input(x, vec):
...
+    
+    If `n = 0`, it returns an identity matrix or batch of identity matrices.
+
+    If `n < 0`, it returns the inverse of each matrix (if invertible) raised to
+    the power of `-n`.
+
+    Args:
+        x (Tensor): A square matrix or a batch of square matrices to be raised
+            to power(n). Its data type should be float or double.
+        n (int): The exponent. It can be any positive, negative integer or zero.
```
Comment (jzhang533) [inline python/paddle/tensor/linalg.py:970]: name参数的说明

File: python/paddle/tensor/linalg.py [hunk 3]
```diff
@@ -941,3 +941,55 @@ def __check_input(x, vec):
...
+
+    Returns:
+        Tensor: The n-th power of matrix (or batch of matrices) x. Its data type is
+            the same with the data type of x.
+
+    Examples:
+        .. code-block:: python
+
+            import paddle
```
Comment (jzhang533) [inline python/paddle/tensor/linalg.py:981]: 示例可以覆盖n>0, n=0, n<0 的情况

File: python/paddle/tensor/linalg.py (hunk 2 above)
Comment (haolin-nju) [reply python/paddle/tensor/linalg.py:970]: Done.

File: python/paddle/tensor/linalg.py (hunk 3 above)
Comment (haolin-nju) [reply python/paddle/tensor/linalg.py:981]: Done.

File: python/paddle/tensor/linalg.py (hunk 1 above)
Comment (haolin-nju) [reply python/paddle/tensor/linalg.py]: Done.

**Review by haolin-nju — COMMENTED** [review]
Resolve comments in matrix_power api.

File: paddle/fluid/operators/matrix_power_op.h [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+#include "paddle/fluid/operators/math/matrix_inverse.h"
+#include "paddle/fluid/platform/for_range.h"
+
+namespace paddle {
+namespace operators {
+
+using Tensor = framework::Tensor;
+
+template <typename T>
+struct EyeFunctor {
```
Comment (JiabinYang) [inline paddle/fluid/operators/matrix_power_op.h]: Maybe another name?

File: paddle/fluid/operators/matrix_power_op.h [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+                         const paddle::framework::ExecutionContext& ctx) {
+  const auto& x_dims = X->dims();
+  const int x_ndim = x_dims.size();
+  T* out_data = Out->mutable_data<T>(ctx.GetPlace());
+
+  auto& dev_ctx = ctx.template device_context<DeviceContext>();
+  platform::ForRange<DeviceContext> for_range(dev_ctx, X->numel());
+
+  if (n == 0) {
+    // Out = I
```
Comment (JiabinYang) [inline paddle/fluid/operators/matrix_power_op.h]: wrong comment?

File: paddle/fluid/operators/matrix_power_op.h [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+
+  if (new_n == 1) {
+    framework::TensorCopy(new_x, ctx.GetPlace(), dev_ctx, Out);
+    return;
+  }
+
+  auto no_trans_desc = math::CreateMatrixDescriptor(x_dims, 0, false);
+
+  if (new_n == 2) {
+    // Out = newX \times newX
```
Comment (JiabinYang) [inline paddle/fluid/operators/matrix_power_op.h]: use `*`

**Review by JiabinYang — DISMISSED** [review]
LGTM

File: paddle/fluid/operators/matrix_power_op.h (hunk 4 above)
Comment (haolin-nju) [reply paddle/fluid/operators/matrix_power_op.h]: This naming method refers to that of [eye_op.h L27](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/eye_op.h#L27) and [cholesky_op.h L88](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/cholesky_op.h#L88). If it's necessary to rename this functor, what about `IdentityMatrixFunctor`?

File: paddle/fluid/operators/matrix_power_op.h (hunk 5 above)
Comment (haolin-nju) [reply paddle/fluid/operators/matrix_power_op.h]: `I` means identity matrix here. However, there are two notations, `I` and `E`, for identity matrix in mathematics. I have no idea which one would paddle prefer. Could you please give me some advice?

**Review by haolin-nju — COMMENTED** [review]
Need more advice about naming method and comment message.

File: paddle/fluid/operators/matrix_power_op.h (hunk 4 above)
Comment (haolin-nju) [reply paddle/fluid/operators/matrix_power_op.h]: Done.

File: paddle/fluid/operators/matrix_power_op.h (hunk 5 above)
Comment (haolin-nju) [reply paddle/fluid/operators/matrix_power_op.h]: Done.

File: paddle/fluid/operators/matrix_power_op.h (hunk 6 above)
Comment (haolin-nju) [reply paddle/fluid/operators/matrix_power_op.h]: Done.

**Review by haolin-nju — COMMENTED** [review]
Three comments have been resolved in the [commit(ce2fa089)](https://github.com/PaddlePaddle/Paddle/pull/34667/commits/ce2fa089ab639c88f66d76a74a6537a34272d0bd)

**Review by JiabinYang — APPROVED** [review]
LGTM

**Review by juncaipeng — APPROVED** [review]
LGTM for changing op_threshold_white_list.py

**Review by dingjiaweiww — APPROVED** [review]

**Review by lanxianghit — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LG API

**Pull request merged**